### DISCUSSION TITLE
Inbuilt legacy module for shared openssl mac universal libs

### DISCUSSION THREAD

**Discussion by aniljadaun** [openssl/openssl#21525] [Q&A]
Is it possible to created inbuilt legacy module with shared openssl dylibs ?

i have created libcrypto.dylib and libssl.dylib with perl Configure no-module ?

Am i missing something ?

**Comment by paulidale** [comment]
Try `no-shared`.

**Reply by aniljadaun** [reply]
Hi @paulidale 
no-shared will create static libs right ? I am looking to create dynamic libs with no separate libs for legacy.

**Reply by aniljadaun** [reply]
Hi @mattcaswell @t8m 

please suggest

**Comment by t8m** [comment]
This is duplicate of your previous question.

**Reply by aniljadaun** [reply]
Hi @t8m @mattcaswell 

This is related but different question.
I created static libs with inbuilt legacy  for windows using 
perl Configure no-module no-shared and it worked fine for me.

But for mac we are using dynamic linking and thus i am using 
perl Configure no-module  but some how legacy provider is not working. So i wanted to know if their is a difference in behaviour of
static and dynamic libs

Is it possible to create dynamic libs with inbuilt legacy module ?

**Reply by aniljadaun** [reply]
@levitte Can you please suggest on this ? Is it possible to produce inbuilt legacy provider in dynamic libcrypto.dylib and libssl.dylib ?

**Reply by t8m** [reply]
> But for mac we are using dynamic linking and thus i am using
> perl Configure no-module but some how legacy provider is not working. So i wanted to know if their is a difference in behaviour of
> static and dynamic libs

It should be sufficient to use no-module and the legacy provider should be built in. It should work fine with dynamically loaded libcrypto/libssl. How do you activate the legacy provider? Does OSSL_PROVIDER_load() return error?

**Reply by aniljadaun** [reply]
Hi @t8m 
   
   OSSL_PROVIDER* leg;
    leg = OSSL_PROVIDER_load(NULL, "legacy");
    if(!leg){
  
       FILE* fptr = NULL;
       if ((fptr = fopen("/Users/anils/log.txt", "a+")) != NULL) {
           fprintf(fptr, "%s \n", "Error logging ::::");
          ERR_print_errors_fp(fptr);
          fclose(fptr);
      }
    }

This is the error i am getting



0016870202000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:341:Global default library context, Algorithm (mdc2 : 98), Properties (<null>)
0016870202000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:341:Global default library context, Algorithm (mdc2 : 98), Properties (<null>)
0016870202000000:error:0308010C:digital envelope routines:inner_evp_generic_fetch:unsupported:crypto/evp/evp_fetch.c:341:Global default library context, Algorithm (mdc2 : 98), Properties (<null>)
0016870202000000:error:12800067:DSO support routines:dlfcn_load:could not load the shared library:crypto/dso/dso_dlfcn.c:118:filename(/Users/anils/openssl/x86_64/dll/release/lib/ossl-modules/legacy.dylib): dlopen(/Users/anils/openssl/x86_64/dll/release/lib/ossl-modules/legacy.dylib, 0x0002): tried: '/Users/anils/openssl/x86_64/dll/release/lib/ossl-modules/legacy.dylib' (no such file), '/usr/local/lib/legacy.dylib' (no such file), '/usr/lib/legacy.dylib' (no such file)
0016870202000000:error:12800067:DSO support routines:DSO_load:could not load the shared library:crypto/dso/dso_lib.c:152:
0016870202000000:error:07880025:common libcrypto routines:provider_init:reason(524325):crypto/provider_core.c:904:name=legacy

**Reply by aniljadaun** [reply]
I am not sure why i see refernce to my local system path...

**Reply by t8m** [reply]
This is weird. Are you sure your testcode is linked (at runtime) with libcrypto.dylib that is built with no-module (and without no-legacy)?

**Reply by aniljadaun** [reply]
Hi @t8m 

Thanks for your help, not sure what i missed. Cleared all libs from my system and recreated libs from step 1 and it is working now. I am not sure what exactly i messed up in the previous try. Good news is that it is working now. 
Thank you so much for your help. I truly appreciate it beyond words.
